sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -146,10 +155,32 @@ case class HashAggregateExec(
   // but the vectorized hashmap can still be switched on for testing and benchmarking purposes.
   private var isVectorizedHashMapEnabled: Boolean = false
 
+  // VisibleForTesting
+  private[sql] lazy val isAdaptivePartialAggregationEnabled = {
+    isPartialAgg && groupingAttributes.nonEmpty && conf.adaptivePartialAggregationThreshold > 0 &&
+      conf.adaptivePartialAggregationThreshold < (1 << conf.fastHashAggregateRowMaxCapacityBit) && {
+      child
+        .collectUntil(p => p.isInstanceOf[WholeStageCodegenExec] ||
+          !p.isInstanceOf[CodegenSupport] ||
+          p.isInstanceOf[LeafExecNode]).forall {
+        case _: ProjectExec | _: FilterExec | _: ColumnarToRowExec => true
+        case _: SerializeFromObjectExec => true
+        case _: InputAdapter => true
+        // HashAggregateExec, ExpandExec, SortMergeJoinExec ...
+        case _ => false
```

Comment:
Why does it need to limit child node types? Is this for a performance reason? If so, after nodes like aggregate, join, and expand, skipping partial aggregate may still benefit performance. Why isn’t adaptivePartialAggregationThreshold good enough?